scripts/util/doc-builder.js

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env node
+
+var path = require('path');
+var fs = require('fs');
+var yuidoc = require('./yuidoc');
+var NodeGit = require('nodegit');
+var util = require('util');
+var fsExtra = require('fs-extra');
+
+function DocBuilder(options) {
+  this.options = util._extend({
+    masterBranchName: 'master',
+    docsBranchName: 'gh-pages',
+    docsDirName: 'docs',
+    versionsDirName: 'versions',
+    srcDirName: 'src',
+    rmSrcDir: false,
+    commitDocs: false,
+    checkoutStrategy: ( NodeGit.Checkout.STRATEGY.FORCE + NodeGit.Checkout.STRATEGY.DONT_WRITE_INDEX ),
+  }, options);
+
+  if (!this.options.themeDir) {
+    this.options.themeDir = path.join(this.options.docsDirName, 'yuidoc', 'theme');
+  }
+
+  this.options.apiDirPath = path.join(this.options.docsDirName, this.options.versionsDirName);
+}
+
+DocBuilder.prototype.ensureDocsBranchExists = function () {
+  console.info('info:', 'Ensuring \'' + this.options.docsBranchName + '\' branch exists');
+  return NodeGit.Repository.open('.').then(repo => {
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(references => {
+      var docsBranchReference;
+      var regex = new RegExp('/' + this.options.docsBranchName + '$'); //ensure reference name ends with the provided branch name
+
+      for (key in references) {
+        if (references[key].name().match(regex)) {
+          docsBranchReference = references[key];
+          if (!references[key].isRemote()) {
+            break;
+          }
+        }
+      };
```

Comment:
IMO it's fine to do this as a single filter operation, and you can use `find` if you only want one:

```js
docsBranch = Object
  .keys(references)
  .map((key) => references[key])
  .find((branch) => branch.name().match(regex) && branch.isRemote());

if (docBranch == null) {
  throw new Error(...)
}
```